### PR TITLE
Add ModelOpt transformer model pruning example for Llama3 model

### PR DESCRIPTION
**Pull request by shengliangxu**
# What does this PR do ?

Add ModelOpt transformer model pruning example for Llama3 model

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nlp/language_modeling/megatron_gpt_prune.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+    model_pruned.save_to(cfg.export.save_path)
+
+    torch.distributed.barrier()
+
+    # TODO: read back and run inference
+
+
+if __name__ == '__main__':
+    main()
```
Comment (kevalmorabia97) [inline nlp/language_modeling/megatron_gpt_prune.py:149]: Cleaned up your code

````python
import modelopt.torch.prune as mtp
import torch.multiprocessing as mp
from datasets import load_dataset
from omegaconf import OmegaConf
from pytorch_lightning.trainer.trainer import Trainer
from tqdm import tqdm

from nemo.collections.nlp.models.language_modeling.megatron_gpt_model import MegatronGPTModel
from nemo.collections.nlp.parts.nlp_overrides import NLPDDPStrategy
from nemo.core.config import hydra_runner
from nemo.utils.model_utils import load_config

mp.set_start_method("spawn", force=True)

"""
Nemo pruning example script.

Please consult examples/nlp/language_modeling/conf/megatron_gpt_prune.yaml config on available pruning arguments,
models supported as well as how to set up data and inference for calibration (with defaults recommended).

Example usage:
```
python examples/nlp/language_modeling/megatron_gpt_prune.py \
    model.restore_from_path=llama3.1-8b-base.nemo \
    model.tensor_model_parallel_size=1 \
    model.pipeline_model_parallel_size=8 \
    trainer.num_nodes=1 \
    trainer.precision=bf16 \
    trainer.devices=8 \
    prune.save_path=llama3.1-8b-base-pruned.nemo \
    prune.ffn_hidden_size=3584 \
    prune.num_attention_heads=8 \
    prune.num_query_groups=4
```
where tensor_model_parallel_size must be 1 because of the current prune API limitation
"""


def get_calib_data_iter(data="cnn_dailymail", batch_size=64, calib_size=512, max_sequence_length=512):
    if data == "wikitext":
        dataset = load_dataset("wikitext", "wikitext-103-v1", split="train")
        text_column = "text"
    elif data == "cnn_dailymail":
        dataset = load_dataset("cnn_dailymail", name="3.0.0", split="train")
        text_column = "article"
    else:
        # Assume a local JSON dataset with a column named "text"
        dataset = load_dataset("json", data_files=data, split="train")
        text_column = "text"
    calib_size = max(min(len(dataset), calib_size), batch_size)
    for i in range(calib_size // batch_size):
        batch = dataset[i * batch_size : (i + 1) * batch_size][text_column]
        for j in range(len(batch)):
            batch[j] = batch[j][:max_sequence_length]
        yield batch


@hydra_runner(config_path="conf", config_name="megatron_gpt_prune")
def main(cfg) -> None:
    # Overwrite model config with the one from the model checkpoint and apply pruning modifications
    model_cfg = load_config(cfg.model.restore_from_path)
    model_cfg.update(cfg.model)
    model_cfg.name = "modelopt"  # Use modelopt transformer spec for pruning

    assert cfg.model.tensor_model_parallel_size == 1, "Pruning currently only supports tensor_model_parallel_size=1"
    assert not cfg.model.sequence_parallel, "Pruning currently does not support sequence parallelism"

    trainer = Trainer(strategy=NLPDDPStrategy(), **cfg.trainer)
    model = MegatronGPTModel.restore_from(
        restore_path=cfg.model.restore_from_path, override_config_path=model_cfg, trainer=trainer
    )

    data_iter = get_calib_data_iter(
        cfg.prune.calib_dataset,
        cfg.inference.batch_size,
        cfg.prune.num_calib_size,
        cfg.inference.max_context_length,
    )
    dataloader = [data for data in data_iter]

    def forward_loop(model):
        # NOTE: Alternatively you can also use `model.forward_bwd_step(data_iter, forward_only=True)`
        # if your model is setup for training.
        model.set_inference_config(OmegaConf.to_container(cfg.inference))
        for i, batch in enumerate(tqdm(dataloader, desc="Calibrating")):
            model.predict_step(batch, i)

    model_pruned, _ = mtp.prune(
        model,
        mode="mcore_gpt_minitron",
        constraints={
            "export_config": {
                k: cfg.prune.get(k)
                for k in ["ffn_hidden_size", "num_attention_heads", "num_query_groups"]
                if cfg.prune.get(k) is not None
            },
        },
        dummy_input=None,  # Not used
        config={"forward_loop": forward_loop},
    )

    model_pruned.save_to(cfg.prune.save_path)


if __name__ == '__main__':
    main()
````

File: nlp/language_modeling/conf/megatron_gpt_prune.yaml [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  num_query_groups: 4 # num_query_groups in the pruned model, num_query_groups // 2
+
+export:
+  decoder_type: llama # gptnext, gpt2, llama
+  inference_tensor_parallel: 1 # Default using 1 TP for inference
+  inference_pipeline_parallel: 1 # Default using 1 PP for inference
+  dtype: ${trainer.precision} # Default precision data type
+  save_path: llama3-8b-base-pruned.nemo # Path where the pruned model will be saved
+  compress: false # Whether save_path should be a tarball or a directory
+  sample_output: true # Whether to run a sample prompt before saving
```
Comment (kevalmorabia97) [inline nlp/language_modeling/conf/megatron_gpt_prune.yaml:47]:
```suggestion
model:
  tensor_model_parallel_size: 1 # can only be 1 for now
  pipeline_model_parallel_size: 1
  sequence_parallel: False # should be disabled during pruning
  restore_from_path: ???

prune:
  calib_dataset: cnn_dailymail # wikitext, cnn_dailymail, or a local dataset
  num_calib_size: 1024 # number of samples used for calibration
  save_path: ??? # Path where the pruned model will be saved
  ffn_hidden_size: null # ffn_hidden_size in the pruned model
  num_attention_heads: null # num_attention_heads in the pruned model
  num_query_groups: null # num_query_groups in the pruned model
```

Comment (shengliangxu) [reply nlp/language_modeling/conf/megatron_gpt_prune.yaml:47]: The values in this file are all default values. So we should provide a working value set. They can be overwritten through command line arguments.

Comment (kevalmorabia97) [reply nlp/language_modeling/conf/megatron_gpt_prune.yaml:47]: Thats the convension in most config files with the exception of the ptq script.
- https://github.com/NVIDIA/NeMo/blob/main/examples/nlp/language_modeling/conf/megatron_llama_distill.yaml
- https://github.com/NVIDIA/NeMo/blob/main/examples/nlp/language_modeling/conf/megatron_llama_config.yaml
- https://github.com/NVIDIA/NeMo/blob/main/examples/nlp/language_modeling/conf/megatron_gpt_config.yaml

I think these paths for loading and saving are always going to changed based on the user and the pruning parameters are also model dependent. Having them default to something might cause unexpected behavior if users don't overwrite them for their model and they wont get any errors as well

Comment (shengliangxu) [reply nlp/language_modeling/conf/megatron_gpt_prune.yaml]: I believe it is fine and much better to have the values set because we provide this setting:

`  restore_from_path: llama3-8b-base.nemo # Nemo file path`

meaning the default config is for this model and we have tested the setting for this particular model. Users can just grab and place the model here and everything should just work.

a ? or a ?? provide no info for the user.

File: examples/nlp/language_modeling/conf/megatron_gpt_prune.yaml [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  activations_checkpoint_granularity: null # 'selective' or 'full'
+  activations_checkpoint_method: null # 'uniform', 'block', not used with 'selective'
+
+prune:
+  decoder_type: ${export.decoder_type} # gptnext, gpt2, llama
+  calib_dataset: cnn_dailymail # wikitext, cnn_dailymail, or a local dataset
+  num_calib_size: 512 # number of samples used for calibration
+  ffn_hidden_size: 3584 # ffn_hidden_size in the pruned model, ffn_hidden_size // 4
+  num_attention_heads: 8 # num_attention_heads in the pruned model, num_attention_heads // 4
+  num_query_groups: 4 # num_query_groups in the pruned model, num_query_groups // 2
```
Comment (kevalmorabia97) [inline examples/nlp/language_modeling/conf/megatron_gpt_prune.yaml:38]:
```suggestion
  num_query_groups: 4 # num_query_groups in the pruned model, num_query_groups // 2
  hidden_size: 2048 # hidden_size in the pruned model, hidden_size // 2
```

File: examples/nlp/language_modeling/megatron_gpt_prune.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+python examples/nlp/language_modeling/megatron_gpt_prune.py \
+    model.restore_from_path=llama3.1-8b-base.nemo \
+    model.tensor_model_parallel_size=1 \
+    model.pipeline_model_parallel_size=8 \
+    trainer.num_nodes=1 \
+    trainer.precision=bf16 \
+    trainer.devices=8 \
+    prune.ffn_hidden_size=3584 \
+    prune.num_attention_heads=8 \
+    prune.num_query_groups=4 \
```
Comment (kevalmorabia97) [inline examples/nlp/language_modeling/megatron_gpt_prune.py:47]:
```suggestion
    prune.num_query_groups=4 \
    prune.hidden_size=2048 \
```

File: examples/nlp/language_modeling/megatron_gpt_prune.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        for i, batch in enumerate(tqdm(dataloader, desc="Calibrating")):
+            model.predict_step(batch, i)
+
+    model_pruned, _ = mtp.prune(
+        model,
+        mode="mcore_gpt_minitron",
+        constraints={
+            "export_config": {
+                k: cfg.prune.get(k)
+                for k in ["ffn_hidden_size", "num_attention_heads", "num_query_groups"]
```
Comment (kevalmorabia97) [inline examples/nlp/language_modeling/megatron_gpt_prune.py:112]:
```suggestion
                for k in ["ffn_hidden_size", "num_attention_heads", "num_query_groups", "hidden_size"]
```

**Comment by shengliangxu** [comment]
The changes has passed testing on a H100x8 system.

**Comment by kevalmorabia97** [comment]
Tested this example on Llama3.1-8B and NeMo Mistral 12B for embedding, attention, mlp pruning. Model can be pruned and correctly restored from output .nemo for further finetuning

**Review by kevalmorabia97 — APPROVED** [review]
This example is tested by me and Saurav M.

File: examples/nlp/language_modeling/conf/megatron_gpt_prune.yaml [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+  tensor_model_parallel_size: 1 # Pruning currently only supports tensor_model_parallel_size=1
+  pipeline_model_parallel_size: 1
+  restore_from_path: llama3.1-8b-base.nemo # Nemo file path
+
+  ## Activation Checkpoint
+  activations_checkpoint_granularity: null # 'selective' or 'full'
+  activations_checkpoint_method: null # 'uniform', 'block', not used with 'selective'
+
+prune:
+  decoder_type: ${export.decoder_type} # gptnext, gpt2, llama
```
Comment (kevalmorabia97) [inline examples/nlp/language_modeling/conf/megatron_gpt_prune.yaml:33]:
```suggestion
```

**Review by kevalmorabia97 — APPROVED** [review]

**Comment by ericharper** [comment]
There's no unit tests or CI tests. Can we add those?

**Review by ko3n1g — CHANGES_REQUESTED** [review]
Please add this test to the final check too (`Nemo_CICD_Test.needs`)

**Review by kevalmorabia97 — APPROVED** [review]

**Review by ko3n1g — COMMENTED** [review]
Can you rebase against upstream main? The CI can’t run if there’s a merge conflict

**Review by kevalmorabia97 — APPROVED** [review]

**Review by ko3n1g — DISMISSED** [review]
CI needs to pass ofc but code-wise it looks good

**Review by kevalmorabia97 — APPROVED** [review]

**Review by ko3n1g — APPROVED** [review]
